### ISSUE TITLE
Typed operations

### ISSUE THREAD

**Issue by markcatley** [pdf-rs/pdf#85]
I am writing a tool that extracts data out of some domain-specific pdfs and am finding it quite cumbersome to deal with the untyped operations.

I am writing an enum version of `pdf::content::Operation`. Is that something you'd be interested in having in-tree or should I create a separate repo and crate for it?

**Comment by s3bk** [comment]
Definitely!

The current state is a lazy hack at best.
You can probably reuse a lot from this code for parsing: https://github.com/pdf-rs/pdf_render/blob/master/render/src/renderstate.rs#L251

I think for location, the pdf repo fits best.

**Comment by s3bk** [comment]
@markcatley Did you get anywhere with this?
I think it would be quite useful to have and I don't mind putting in a few hours to do it, if you didn't write it yet.

**Comment by markcatley** [comment]
@s3bk it's been a crazy month! I've extracted the code I used into a repo/crate for you to have a look at. I've run the binary on all of the PDFs I have sitting around on my laptop. A few panic in pdf-rs code the rest all parse into the typed operation except where noted in the readme. I'll open some issues about the panics when I get a chance.

https://github.com/markcatley/typed-pdf/ & https://crates.io/crates/typed-pdf

Let me know what you think.

**Comment by s3bk** [comment]
I see you went the way and translated them one to one.
My approach (implemented yesterday) is https://github.com/pdf-rs/pdf/blob/1aaca192b945f5be5f98e11147aaa5ad3b37f5a1/pdf/src/content.rs#L642-736 mostly motivated by writing the renderer where it helps to merge similar operators like `B`, `b`, `B*` and `b*`.

**Comment by markcatley** [comment]
I do like yours better. No great surprise given you're likely much more experienced with PDF handling than me. I'll have a go with it soon. In reading, the only feedback I had was: The stroke booleans are a bit hard to follow. I'd probably change them to an `enum OperationType { Stroking, NonStroking }` or similar.

**Comment by s3bk** [comment]
I managed to remove the `bool`s and instead generate an `Op::Close` instead:
https://github.com/pdf-rs/pdf/commit/9780a38592e6c7de567afc1d5572b0db44bc5798

`Op::Fill`, `Op::Stroke` and `Op::FillAndStroke` will probably stay (or merge into a `Op::Draw(DrawMode))` for performance reasons.
```rust
pub enum DrawMode {
    Fill,
    Stroke,
    FillThenStroke
}
```

**Comment by s3bk** [comment]
just so you know: I have hit `BI` and the corresponding `unimplemented()!`. The implementation is progressing slowly, but it is quite annoying.

**Comment by markcatley** [comment]
Sounds good, to be honest. I haven't written much code with either of those features so take my feedback with a grain of salt.

I've updated my utility which mostly about extracting text from PDFs. I noticed the following:
- `TextMatrix` and `TextPosition` are a little hard for me to follow based on how they're named, I have to go back and look at the spec to remember what they mean. It might be better to use `SetTextMatrix` and `MoveTextPosition`.

I also need to run the pdfs on my disk through the parsing algorithm and check that they parse correctly (with the exception of the unimplemented features of course). I'll get to that when I can.
I also will try to have a look through the naming of the rest of the enum variants and see if any of the names could be improved. However, it's a bit harder with the ops I haven't used much before.
